File: 2022/AoC.py
import os
import argparse

TEST = False
DAY = None

exec_part1 = True
exec_part2 = True

def tprint(msg):
    if TEST:
        print(msg)

def get_input(day=None):
    if day is None:
        if DAY:
            day = DAY
        else:
            raise Exception("Day has not been set")
    filename = "inputs/day{}{}.input".format(day,".test" if TEST else "")
    if os.path.exists(filename):
        with open(filename,'r') as fp:
            return fp.read()
    raise Exception("Cannot find input file for day {}".format(day))

def set_day(day):
    global DAY
    DAY = day

def get_default_argparse():
    desc = "Advent of Code solutions"
    if DAY:
        desc = desc + " -- Day {}".format(DAY)
    return argparse.ArgumentParser(description=desc)

def parse_args(parser=None):
    global TEST, exec_part1, exec_part2
    if parser is None:
        parser = get_default_argparse()
    parser.add_argument("-t","--test",action="store_true")
    parser.add_argument("--part1",action="store_true")
    parser.add_argument("--part2",action="store_true")
    args = parser.parse_args()

    if args.test:
        TEST = True
    if args.part1 or args.part2:
        exec_part1 = args.part1
        exec_part2 = args.part2
    return args

File: 2024/aoc.lua
if AOC ~= nil then
	return AOC
end
AOC = {}
AOC.test_mode = false
AOC.debug = false

local context_registered = false
local this_context = {}

function AOC.dprint(msg)
	if AOC.debug then
		print(msg)
	end
end

function AOC.tprint(msg)
	if AOC.test_mode or AOC.debug then
		print(msg)
	end
end

function AOC.get_input_file_name(custom)
	s = "inputs/" .. this_context.short_name
	if AOC.test_mode then
		s = s .. ".test"
	end
	if custom then
		s = s .. "." .. custom
	end
	s = s .. ".input"
	return s
end

function AOC.get_input_file(custom)
	fname = AOC.get_input_file_name(custom)
	return io.open(fname,'r')
end

function AOC.register(short_name, long_name, part1, part2)
	this_context.short_name = short_name
	this_context.long_name = long_name
	this_context.part1 = part1
	this_context.part2 = part2
	context_registered = true
end

function AOC.start()
	if context_registered == false then
		print("ERROR: No AOC module has been registered yet")
		os.exit(10)
	end

	local argparse = require("argparse")
	local parser = argparse(this_context.short_name,"Advent of Code -- " .. this_context.long_name)
	parser:flag("-1 --part1","Execute Part 1")
	parser:flag("-2 --part2","Execute Part 2")
	parser:flag("-t --test", "Enable test mode")
	parser:flag("-d --debug", "Enable debug output (includes test mode output)")
	parser:option("-i --input","Use a custom input file")

	local args = parser:parse()
	if args.test then
		AOC.test_mode = true
		print("** TESTING **")
	end

	if args.debug then
		AOC.debug = true
	end

	exec_p1 = args.part1 and true or false
	exec_p2 = args.part2 and true or false
	if exec_p1 == false and exec_p2 == false then
		exec_p1 = true
		exec_p2 = true
	end

	print("== " .. this_context.long_name .. " ==")
	local fname = args.input or AOC.get_input_file_name()
	if args.input then
		fname = args.input
	end
	local f = io.open(fname,'r')
	if not f then
		print("ERROR: Cannot open input file '" .. fname .. "'")
		os.exit(8)
	end
	io.close(f)
	if exec_p1 then
		print("-- Part 1 --")
		f = io.open(fname,'r')
		this_context.part1(f)
		io.close(f)
	end

	if exec_p2 then
		print("-- Part 2 --")
		f = io.open(fname,'r')
		this_context.part2(f)
		io.close(f)
	end
end

return AOC
